File: src/Exchange/Main/MatchingEngine/Order.java
package Exchange.Main.MatchingEngine;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;

public class Order {
    private long TimeStamp;
    private OrderSide oSide; // Buy or Sell
    private OrderClass oClass; // Market or Limit
    private Integer Quantity;
    private BigDecimal Price;
    private BigInteger orderID;
    private BigInteger customerID;
    private BigInteger instrumentID;
    private Order PreviousOrder;
    private Order NextOrder;
    private Limit parentLimit;

    public Order(OrderSide oSide, OrderClass oClass, Integer quantity,
                 BigDecimal price, BigInteger orderID, BigInteger customerID, BigInteger instrumentID) {
        TimeStamp = System.nanoTime();
        this.oSide = oSide;
        this.oClass = oClass;
        Quantity = quantity;
        Price = price.setScale(3, RoundingMode.HALF_UP); // Rounds to the nearest thousandth
        this.orderID = orderID;
        this.customerID = customerID;
        this.PreviousOrder = null;
        this.NextOrder = null;
        this.instrumentID = instrumentID;
    }

    public Limit getParentLimit() {
        return parentLimit;
    }

    public void setParentLimit(Limit parentLimit) {
        this.parentLimit = parentLimit;
    }

    public long getTimeStamp() {
        return TimeStamp;
    }

    public OrderSide getoSide() {
        return oSide;
    }

    public OrderClass getoClass() {
        return oClass;
    }

    public Integer getQuantity() {
        return Quantity;
    }

    // Always rounds so 20.1 is actually 20.100
    public BigDecimal getPrice() {
        return Price;
    }

    public BigInteger getOrderID() {
        return orderID;
    }

    public BigInteger getCustomerID() {
        return customerID;
    }

    public Order getPreviousOrder() {
        return PreviousOrder;
    }

    public void setPreviousOrder(Order previousOrder) {
        PreviousOrder = previousOrder;
    }

    public Order getNextOrder() {
        return NextOrder;
    }

    public void setNextOrder(Order nextOrder) {
        NextOrder = nextOrder;
    }

    @Override
    public String toString() {
        return "Order{" +
                "TimeStamp=" + TimeStamp +
                ", oSide=" + oSide +
                ", oClass=" + oClass +
                ", Quantity=" + Quantity +
                ", Price=" + Price +
                ", orderID=" + orderID +
                ", customerID=" + customerID +
                ", instrumentID=" + instrumentID +
                '}';
    }

    public void setOrderID(BigInteger orderID) {
        this.orderID = orderID;
    }

    public void setQuantity(Integer quantity) {
        Quantity = quantity;
    }

    public void updateQuantity(Integer newQuantity, long TimeStamp) {
        if ((newQuantity > this.Quantity) && (this.parentLimit.getTailOrder() != this)) {
            // Move order to the end of the list. i.e. loses time priority
            this.parentLimit.movetoTail(this);
            this.TimeStamp = TimeStamp;
        }
        parentLimit.setVolume(parentLimit.getVolume()-(this.Quantity- newQuantity));
        this.Quantity = newQuantity;
    }

    public BigInteger getInstrumentID() {
        return instrumentID;
    }
}

File: src/GAMFinancial-WebApp/GAMFinancial-WebApp/Models/Order.cs
using System.ComponentModel.DataAnnotations;
using System.Numerics;

namespace GAMFinancial_WebApp.Models
{
    public class Order
    {
        public Order(OrderSide oSide, OrderClass oClass, int quantity,
                 decimal price, BigInteger orderID, BigInteger customerID, BigInteger instrumentID, bool isNew)
        {
            TimeStamp = DateTime.UtcNow.Ticks * 100; // Time in nanosconds
            this.OrderSide = oSide;
            this.OrderClass = oClass;
            this.Quantity = quantity;
            this.Price = Math.Round(price, 3); // Rounds to the nearest thousandth
            this.OrderID = orderID;
            this.CustomerID = customerID;
            this.PreviousOrder = null;
            this.NextOrder = null;
            this.InstrumentID = instrumentID;
            this.IsNew=isNew;
        }

        [Key]
        public BigInteger OrderID { get; set; }

        public bool IsNew { get; set; }

        public long TimeStamp { get; set; }

        public OrderSide OrderSide { get; }

        public OrderClass OrderClass { get; }

        public int Quantity { get; set; }

        public decimal Price { get; }
        

        public BigInteger CustomerID { get; set; }

        public BigInteger InstrumentID { get; set; }

        public Order? PreviousOrder { get; set; }

        public Order? NextOrder { get; set; }

        public Limit? ParentLimit { get; set; }

        public void updateQuantity(int newQuantity, long TimeStamp)
        {
            if ((newQuantity > this.Quantity) && (this.ParentLimit.TailOrder != this))
            {
                // Move order to the end of the list. i.e. loses time priority
                this.ParentLimit.movetoTail(this);
                this.TimeStamp = TimeStamp;
            }
            this.ParentLimit.Volume = this.ParentLimit.Volume-(this.Quantity- newQuantity);
            this.Quantity = newQuantity;
        }
    }
}
